File: 2707. Extra Characters in a String/minExtraChar.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
const minExtraChar = (s, dictionary) => {
  const l = s.length;
  const dictionarySet = new Set(dictionary);
  const dp = new Array(l + 1).fill(0);

  for (let i = 1; i <= l; i++) {
    dp[i] = dp[i - 1] + 1;
    for (let j = 0; j < i; j++) {
      if (dictionarySet.has(s.slice(j, i))) {
        dp[i] = Math.min(dp[i], dp[j]);
      }
    }
  }

  return dp[l];
};


const s = "eleetscode";
const dictionary = ["leet", "code", "leetcode"];
console.log('minExtraChar: ', minExtraChar(s, dictionary)); // 0

File: 2707. Extra Characters in a String/minExtraChar.py
from typing import List

class Solution:
  def minExtraChar(self, s: str, dictionary: List[str]) -> int:
    l = len(s)
    dictionarySet = set(dictionary)
    dp = [0] * (l + 1)
    
    for start in range(l - 1, -1, -1):
      dp[start] = dp[start + 1] + 1
      for end in range(start + 1, l + 1):
        current = s[start:end]
        if current in dictionarySet:
          dp[start] = min(dp[start], dp[end])
          
    return dp[0]


s = "eleetscode"
dictionary = ["leet","code","leetcode"]
solution = Solution()
print('minExtraChar: ', solution.minExtraChar(s, dictionary)) # 0
